File: packages/native/router-swift/Sources/PyreonRouter/RouterView.swift
// RouterView — the SwiftUI counterpart to @pyreon/router's
// `<RouterView />` component.
//
// Phase A4 (route-table dispatcher): RouterView renders the matched
// component when the path matches an entry in `routes`.
// Phase A4.5 (nested routes): RouterView reads `@Environment(\.routerDepth)`
// to know which level of the matched chain to render. Top-of-tree
// `<RouterView />` has depth 0 (renders the outermost layout / leaf).
// Nested `<RouterView />` inside a layout body reads depth 1 (renders
// the child of that layout). Depth increments AUTOMATICALLY: RouterView
// wraps the invoked component with `.environment(\.routerDepth, depth + 1)`.
// Phase A6 (wildcard-404): when no chain matches AND the router has a
// `notFoundComponent`, render that as the depth-0 fallback.
//
// Backward-compat: if the active router has NO routes AND no
// notFoundComponent (pre-A4 shape, where the host wires its own
// `.navigationDestination(for:)`), RouterView emits `EmptyView()`.

import SwiftUI

/// Active-route view — Phase A4 + A4.5 + A6.
///
/// Render-time decision table:
///   - no router in scope            → `EmptyView()` (defensive — impossible
///     inside a properly-wired `RouterProvider`)
///   - chain matches + depth in range → render `chain[depth].route.component()`
///     with `.environment(\.routerDepth, depth + 1)` so descendant
///     `<RouterView />` instances inside the rendered component pick
///     up the next chain entry automatically.
///   - chain matches + depth out-of-range → render `EmptyView()` (no
///     more chain levels to descend into; the common case for a leaf
///     at the bottom that doesn't include its own `<RouterView />`).
///   - no chain match + notFound set → render `notFoundComponent` (A6)
///   - no chain match + notFound nil → `EmptyView()` (pre-A4 fallback;
///     host's `.navigationDestination(for:)` handles routing)
@available(iOS 17.0, macOS 14.0, *)
public struct RouterView: View {
    @Environment(\.pyreonRouter) private var router: PyreonRouter?
    @Environment(\.routerDepth) private var depth: Int

    public init() {}

    public var body: some View {
        // Reading routes + currentPath via resolveCurrentChain triggers
        // SwiftUI's observation: navigation OR routes-table mutation
        // causes this body to re-evaluate, swapping components.
        if let router, let chain = router.resolveCurrentChain() {
            if depth < chain.count {
                // Phase A4.5: invoke this depth's component and
                // increment routerDepth in its environment so any
                // nested `<RouterView />` inside renders chain[depth+1].
                chain[depth].route.component()
                    .environment(\.routerDepth, depth + 1)
            } else {
                // Depth out of chain — common case for a leaf at the
                // bottom that doesn't include its own <RouterView />.
                EmptyView()
            }
        } else if let router, let notFound = router.notFoundComponent {
            // Phase A6: wildcard-404 catch-all. No route chain matched
            // currentPath; the app configured a notFoundComponent
            // (mirrors the web router's `'*'` / `(.*)` wildcard).
            // Always rendered at depth 0 — 404 isn't a chain.
            notFound()
        } else {
            // Pre-A4 fallback: host's `.navigationDestination(for:)`
            // handles routing externally. Don't paint anything here.
            EmptyView()
        }
    }
}

File: packages/native/router-kotlin/src/main/kotlin/com/pyreon/router/RouterView.kt
// RouterView — Compose counterpart to @pyreon/router's
// `<RouterView />` component.
//
// Phase A4 (native readiness audit, 2026-06 — closes CRIT-2 partial):
// RouterView now renders the route table dispatcher's matched component
// instead of the pre-A4 no-op. The router's `resolveCurrentRoute()`
// walks `routes` in declaration order, picks the first match, and
// returns the matching record. RouterView pulls the Composable from
// that record and invokes it.
//
// Backward-compat: if the active router has NO routes configured (the
// pre-A4 shape where the host writes a `when (router.currentPath)`
// dispatcher in its own code), RouterView emits nothing — the host's
// branching is the source of truth.
//
// Out of scope (separate follow-up PRs):
//   - Nested-route depth indexing (`RouteRecord.children` + per-depth
//     `<RouterView />`) — A4.5
//   - Per-route `beforeEnter` guards wired to navigation — A5
//   - Wildcard-404 catch-all + `notFoundComponent` — A6

package com.pyreon.router

import androidx.compose.runtime.Composable

/**
 * Active-route view — Phase A4.
 *
 * Looks up [PyreonRouter] from the CompositionLocal, calls
 * [PyreonRouter.resolveCurrentRoute] against the route table, and
 * invokes the matched component. Falls through to no-op when:
 *   - no [PyreonRouter] is in scope (impossible inside a properly-wired
 *     [RouterProvider]; defensive default)
 *   - the router has no `routes` configured (backward-compat for
 *     pre-A4 apps using manual `when (router.currentPath)` dispatch)
 *   - no route matches the current path (wildcard-404 is A6 work)
 *
 * Usage (with route table, A4 idiom):
 * ```kotlin
 * @Composable
 * fun TodoApp() {
 *     val router = remember {
 *         PyreonRouter(routes = listOf(
 *             RouteRecord("/") { HomePage() },
 *             RouteRecord("/users/:id") { UserPage() },
 *             RouteRecord("/about") { AboutPage() },
 *         ))
 *     }
 *     RouterProvider(router) { RouterView() }
 * }
 * ```
 *
 * Usage (pre-A4 manual dispatch, still supported via no-routes fallback):
 * ```kotlin
 * @Composable
 * fun TodoApp() {
 *     val router = remember { PyreonRouter() }
 *     RouterProvider(router) {
 *         when (router.currentPath) {
 *             "/about" -> AboutPage()
 *             else -> HomePage()
 *         }
 *     }
 * }
 * ```
 */
@Composable
public fun RouterView() {
    val router = LocalPyreonRouter.current ?: return
    // Reading the routes signal + currentPath inside this Composable
    // means a navigation OR a routes-table mutation triggers
    // recomposition. The route's component is then invoked fresh.
    val resolved = router.resolveCurrentRoute() ?: return
    resolved.first.component()
}
